rfc/0018-certificate-handling.md

```diff
@@ -0,0 +1,259 @@
+and then we outline the different actions that have to be done.
+
+### Kubewarden Fresh Install
+
+The Kubewarden stack has just been installed. The following actions have to be done:
+
+- Generate Certificate Authority
+- Generate certificate for `kubewarden-controller` and sign it with our CA
+- Start the Kubewarden controller HTTPS server using this certificate
+- Register the webhook against the Kubernetes API, using the certificate of our CA inside of
+  `clientConfig.caBundle`
+
+### New Policy Server Defined
+
+A new `PolicyServer` is defined, the following actions have to be done:
+
+- Generate new certificate, sign it with our CA
+- Create Policy Server Deployment, ensure the certificate is mounted inside of the Pods
+
+This is already done by our current code (Kubewarden 1.6.0).
+
+### New `AdmissionPolicy`/`ClusterAdmissionPolicy` is defined
+
+A new policy is defined inside of the cluster, ensure the following action are done:
+
+- Register the webhook against the Kubernetes API, use our CA inside of `clientConfig.caBundle`
+
+> **Note:** it doesn't matter which Policy Server is going to host the policy. All the Policy Server
+> instances use a certificate that is signed by our CA. Hence we always put our CA
+> inside of the `clientConfig.caBundle`.
+
+This is already done by our current code (Kubewarden 1.6.0).
+
+### Policy Server Certificate Renewed
+
+From time to time we need to renew the certificate issued to a Policy Server instance. The main reason
+to perform this operation is to ensure the Policy Server is not using an expired certificate.
+
+To renew a Policy Server certificate the following actions have to be done:
+
+- Generate new certificate, sign it with our CA
+- Force a rollout of the Policy Server deployment. This will ensure the HTTPS server uses the new certificate
+
+> **Note:** there's no need to touch the webhook configuration objects related to the policies
+> hosted by the Policy Server. The new certificate is still signed by the same CA, hence nothing has
+> to be changed.
+
+### Internal CA changes
+
+The internal CA managed by Kubewarden might be recreated (for example, to avoid its expiration). This is the
+most disruptive event that can happen, especially once some policies are deployed.
+
+Once the new CA is generated, the following actions have to be performed:
+
+- For each `PolicyServer` defined:
+  - Generate a new certificate, sign it with the new CA
+  - Force a rollout of the PolicyServer Deployment
+- Kubewarden Controller:
+  - Generate a new certificate, sign it with the new CA
+  - Restart the controller to ensure its HTTPS endpoint uses the new certificate
+- For each webhook configuration managed by Kubewarden (meaning all the policies deployed, plus the kubewarden-controller):
+  - Update the `clientConfig.caBundle`: ensure it contains the certificate of the new CA
+
+This change is massive and could lead to connection errors for as long as the PolicyServer are using a certificate
```

Comment:
Had the thought of configuring 2 caBundle on the WebhookClientConfig to perform the migration, but it only accepts 1 caBundle.

Also I wonder if we need an ordered approach here, first rotating the kubewarden-controller cert with the new CA, then, once the controller detects that its cert comes from the new CA, the reconciliation loops for webhook confs and PolicyServers. But I'm ok punting on this.